app/controllers/books_controller.rb

```diff
@@ -0,0 +1,51 @@
+# frozen_string_literal: true
+
+class BooksController < ApplicationController
+  before_action :find_single_book, only: %i[show edit update destroy]
+
+  def index
+    @books = Book.all
+  end
+
+  def show
+    @reviews = Review.where(book_id: params[:id])
+  end
+
+  def new
+    @book = Book.new
+  end
+
+  def create
+    @book = Book.new(book_params)
+    raise ActiveRecord::RecordInvalid, @book unless @book.valid?
+    raise ActiveRecord::RecordNotSaved, @book unless @book.save
+
```

Comment:
Normally, `ActiveRecord::RecordInvalid` will be automatically raised by `#save!` if the record is not valid, and many other errors will be raised by `#save!` as well. That means we don't have to manually raise them ourselves.

So we have 2 ways to handle the error cases

**First**: use `#save!` and rescue the errors (Seems like you already rescue them in `ApplicationController`)

```ruby
def create
  @book = Book.new(book_params)
  @book.save!
end
```
**Second**: use `#save` and handle error by ourselves (because `#save` does not raise any errors)

```ruby
def create
  @book = Book.new(book_params)
  if @book.save
    # redirect success
  else
    flash[:aleart] = @book.errors.full_messages
    # redirect failed
  end
end
```

Reply:
Thx krubbb :)
thats crystal clear.